### PR TITLE
[coverage] Automatic merger for LLVM profile data

### PR DESCRIPTION
**Pull request by harlanhaskins**
When `build-script` is given `--swift-analyze-code-coverage merged`, this will fire off this worker which responds to tests and merges generated profile data into a final coverage file in the test results folder.

### CODE REVIEW

File: tools/SourceKit/tools/sourcekitd-test/CMakeLists.txt [hunk 1]
```diff
@@ -23,9 +23,9 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
     PROPERTIES
     LINK_FLAGS "-Wl,-rpath -Wl,@executable_path/../lib")
```
Comment (gribozavr) [inline tools/SourceKit/tools/sourcekitd-test/CMakeLists.txt]: I think if you change this to append to `LINK_FLAGS`, then you won't need to duplicate appending `-fprofile-instr-generate` below -- it will come from AddSwift.cmake.

File: utils/build-script-impl [hunk 2]
```diff
@@ -71,7 +71,7 @@ KNOWN_SETTINGS=(
     llvm-enable-assertions      "1"              "enable assertions in LLVM and Clang"
     swift-build-type            "Debug"          "the CMake build variant for Swift"
     swift-enable-assertions     "1"              "enable assertions in Swift"
-    swift-analyze-code-coverage "0"              "enable code coverage analysis in Swift"
+    swift-analyze-code-coverage "default"        "enable code coverage analysis in Swift (set to \`merged\` to merge profdata files and remove intermediaries)"
```
Comment (gribozavr) [inline utils/build-script-impl]: What are the possible option values?  What does `default` mean?  Could we just use an explicit value, like `none`?

**Comment by gribozavr** [comment]
I also don't see a definition of `SWIFT_ANALYZE_CODE_COVERAGE` in the top-level CMakeLists.txt, could you add it?

File: utils/build-script-impl (hunk 2 above)
Comment (harlanhaskins) [inline utils/build-script-impl]: The possible values are:

`utils/build-script --swift-analyze-code-coverage` -> `default`
`utils/build-script` -> `none`
`utils/build-script --swift-analyze-code-coverage merged` -> `merged`

Comment (gribozavr) [inline utils/build-script-impl]: ... so let's document them here :)

Comment (gribozavr) [inline utils/build-script-impl]: And would you mind renaming `default` to `not-merged`?

Comment (harlanhaskins) [inline utils/build-script-impl]: No problem! :+1:

File: tools/SourceKit/CMakeLists.txt [hunk 3]
```diff
@@ -239,9 +239,9 @@ macro(add_sourcekit_executable name)
         PROPERTIES
         LINK_FLAGS "-Wl,-exported_symbol,_main")
 
-      if(SWIFT_ANALYZE_CODE_COVERAGE)
-        set_property(TARGET "${name}" APPEND_STRING PROPERTY
-            LINK_FLAGS " -fprofile-instr-generate=swift-%p.profraw -fcoverage-mapping")
+      if(NOT SWIFT_ANALYZE_CODE_COVERAGE STREQUAL "NONE")
```
Comment (jrose-apple) [inline tools/SourceKit/CMakeLists.txt]: Please use the old test.

File: tools/SourceKit/tools/complete-test/CMakeLists.txt [hunk 4]
```diff
@@ -15,9 +15,9 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
     PROPERTIES
     LINK_FLAGS "-Wl,-rpath -Wl,@executable_path/../lib")
 
-  if(SWIFT_ANALYZE_CODE_COVERAGE)
+  if(NOT SWIFT_ANALYZE_CODE_COVERAGE STREQUAL "NONE")
```
Comment (jrose-apple) [inline tools/SourceKit/tools/complete-test/CMakeLists.txt]: Please use the old test.

File: tools/SourceKit/tools/sourcekitd-repl/CMakeLists.txt [hunk 5]
```diff
@@ -9,9 +9,9 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
     PROPERTIES
     LINK_FLAGS "-Wl,-rpath -Wl,@executable_path/../lib")
 
-  if(SWIFT_ANALYZE_CODE_COVERAGE)
+if(NOT SWIFT_ANALYZE_CODE_COVERAGE STREQUAL "NONE")
```
Comment (jrose-apple) [inline tools/SourceKit/tools/sourcekitd-repl/CMakeLists.txt]: Please use the old test.

File: CMakeLists.txt [hunk 6]
```diff
@@ -58,6 +58,10 @@ option(SWIFT_INCLUDE_DOCS
     "Create targets for building docs."
     TRUE)
 
+option(SWIFT_ANALYZE_CODE_COVERAGE
```
Comment (jrose-apple) [inline CMakeLists.txt]: This should use FALSE or NO as the default. See the CMake docs for [if](https://cmake.org/cmake/help/v3.0/command/if.html).

Comment (harlanhaskins) [inline CMakeLists.txt]: It's no longer a boolean flag -- it's one of three possibilities, `none`, `not-merged`, or `merged`.

Comment (jrose-apple) [inline CMakeLists.txt]: Ugh, I just scrolled down far enough to see that this has modes now. Can you rephrase the description to make that clear?

File: utils/profdata_merge_worker.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        calls llvm-cov merge with up to 10 filenames, plus the current
+        in-progress merge."""
+        if not self.filename_buffer:
+            self.report("no files to merge...")
+            return
+        if os.path.exists(self.profdata_path):
+            os.rename(self.profdata_path, self.profdata_tmp_path)
+            self.filename_buffer.append(self.profdata_tmp_path)
+        cleaned_files = ' '.join(pipes.quote(f) for f in self.filename_buffer)
+        llvm_cmd = ("xcrun llvm-profdata merge -o %s %s"
```
Comment (jrose-apple) [inline utils/profdata_merge_worker.py]: Sadly this may not be the version of llvm-profdata that goes with the clang++ we used to build swiftc. I don't think you need to fix this, but can you put a note in a comment here?

File: CMakeLists.txt (hunk 6 above)
Comment (harlanhaskins) [inline CMakeLists.txt]: Yep, will do.

Comment (harlanhaskins) [inline CMakeLists.txt]: Though maybe we should use FALSE as the `none` case.
So the values are `FALSE`, `NOT-MERGED`, and `MERGED`?

File: utils/profdata_merge_worker.py (hunk 7 above)
Comment (harlanhaskins) [inline utils/profdata_merge_worker.py]: This can likely be provided by CMake, right? If that's the case, I'll add an argument to this script with the absolute path to llvm-profdata. But a TODO will do for now.

File: CMakeLists.txt (hunk 6 above)
Comment (jrose-apple) [inline CMakeLists.txt]: I think `OFF` would be the best of the various CMake names for false, but we don't use that anywhere else, so I'll let you decide. I just want to make sure it's clear that `TRUE` isn't valid here!

Comment (harlanhaskins) [inline CMakeLists.txt]: I think `OFF` makes sense, but for consistency's sake I think we should stick with `FALSE`. I'll default it to that and make sure it propagates as `FALSE` through `build-script`

File: utils/profdata_merge_worker.py (hunk 7 above)
Comment (jrose-apple) [inline utils/profdata_merge_worker.py]: Yeah, you could certainly check CMake's clang++ and see if there's an llvm-profdata next to it, and/or allow it to be specified explicitly in the CMake configuration. But I wouldn't spend too much time on it right now; it's easy to do later. I just didn't want it to break for someone and then them having to spend hours figuring out why.

File: CMakeLists.txt (hunk 6 above)
Comment (harlanhaskins) [inline CMakeLists.txt]: However, we still won't be able to revert to `if(SWIFT_ANALYZE_CODE_COVERAGE)` because it'll default to `FALSE` for all the possible cases.

Comment (jrose-apple) [inline CMakeLists.txt]: Huh? "`if (<variable>)`: True if the variable is defined to a value that is not a false constant".

Comment (harlanhaskins) [inline CMakeLists.txt]: Oh! Must've misread that, then. :+1:

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by shahmishal** [comment]
@swift-ci Please test

File: CMakeLists.txt (hunk 6 above)
Comment (gribozavr) [inline CMakeLists.txt]: CMake has a way to specify stringly-typed variables with multiple allowed values, see line 16.

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -196,11 +197,25 @@ if(PYTHONINTERP_FOUND)
             "${CMAKE_CURRENT_SOURCE_DIR}/../validation-test/lit.site.cfg.in"
             "${validation_test_bin_dir}/lit.site.cfg"
             "validation-test${VARIANT_SUFFIX}.lit.site.cfg")
+        set(profdata_merge_worker
+            "${CMAKE_CURRENT_SOURCE_DIR}/../utils/profdata_merge_worker.py")
+
+        if(SWIFT_ANALYZE_CODE_COVERAGE STREQUAL "MERGED")
+            set(command_profdata_merge_start
```
Comment (gribozavr) [inline test/CMakeLists.txt]: 2 spaces of indentation.

File: test/CMakeLists.txt [hunk 9]
```diff
@@ -196,11 +197,25 @@ if(PYTHONINTERP_FOUND)
...
+                COMMAND ${PYTHON_EXECUTABLE} ${profdata_merge_worker} start -o ${swift_test_results_dir})
```
Comment (gribozavr) [inline test/CMakeLists.txt]: Please quote variables.

File: test/lit.cfg [hunk 10]
```diff
@@ -129,6 +132,43 @@ if config.test_exec_root is None:
 
 ###
 
+class SwiftTest(lit.formats.ShTest):
+    def __init__(self, coverage_mode=None, execute_external=True):
+        lit.formats.ShTest.__init__(self, execute_external)
```
Comment (gribozavr) [inline test/lit.cfg]: Is the keyword arg required here?

File: test/lit.cfg [hunk 11]
```diff
@@ -138,7 +178,7 @@ if platform.system() == 'Darwin':
     config.environment['TOOLCHAINS'] = 'default'
 
 # testFormat: The test format to use to interpret tests.
-config.test_format = lit.formats.ShTest(execute_external=True)
+config.test_format = SwiftTest(coverage_mode=config.coverage_mode.lower())
```
Comment (gribozavr) [inline test/lit.cfg]: s/.lower()// and adjust downstream.

File: tools/SourceKit/CMakeLists.txt [hunk 12]
```diff
@@ -239,9 +239,9 @@ macro(add_sourcekit_executable name)
         PROPERTIES
         LINK_FLAGS "-Wl,-exported_symbol,_main")
 
-      if(SWIFT_ANALYZE_CODE_COVERAGE)
-        set_property(TARGET "${name}" APPEND_STRING PROPERTY
-            LINK_FLAGS " -fprofile-instr-generate=swift-%p.profraw -fcoverage-mapping")
+    if(SWIFT_ANALYZE_CODE_COVERAGE)
+          set_property(TARGET "${name}" APPEND_STRING PROPERTY
```
Comment (gribozavr) [inline tools/SourceKit/CMakeLists.txt]: 2 spaces.

File: tools/SourceKit/CMakeLists.txt [hunk 13]
```diff
@@ -239,9 +239,9 @@ macro(add_sourcekit_executable name)
         PROPERTIES
```
Comment (gribozavr) [inline tools/SourceKit/CMakeLists.txt]: Could you change this invocation to be `set_property(TARGET "..." APPEND_STRING PROPERTY LINK_FLAGS)`?  Then you won't need to duplicate the code coverage flags below?

File: utils/build-script [hunk 14]
```diff
@@ -512,11 +512,13 @@ also build for Apple watchos, but disallow tests that require an watchOS device"
...
-        help="enable code coverage analysis in Swift",
-        action="store_const",
-        const=True,
-        dest="swift_analyze_code_coverage",
-        default=False)
+        help="""enable code coverage analysis in Swift (set to `merged` to merge
+and remove intermediary profdata files)""",
+        const="not-merged",
+        choices=["false", "not-merged", "merged"],
+        nargs='?',
```
Comment (gribozavr) [inline utils/build-script]: Could we always require the option value?  This is an expert-only option and adding subtle implicit behavior only adds debugging time.

**Comment by harlanhaskins** [comment]
@modocache If you get a chance, would you mind reviewing profdata_merge_worker.py? Thanks!

**Comment by modocache** [comment]
Sure thing! @practicalswift is also probably good for some Python opinions. :wink:

File: test/lit.cfg [hunk 15]
```diff
@@ -129,6 +132,43 @@ if config.test_exec_root is None:
 
 ###
 
+class SwiftTest(lit.formats.ShTest):
+    def __init__(self, coverage_mode=None, execute_external=True):
+        lit.formats.ShTest.__init__(self, execute_external=execute_external)
```
Comment (modocache) [inline test/lit.cfg]: I think it's preferable to use `super` here, as in:

```python
super(SwiftTest, self).__init__(execute_external=excute_external)
```

File: test/lit.cfg [hunk 16]
```diff
@@ -129,6 +132,43 @@ if config.test_exec_root is None:
 
 ###
 
+class SwiftTest(lit.formats.ShTest):
+    def __init__(self, coverage_mode=None, execute_external=True):
```
Comment (modocache) [inline test/lit.cfg]: I'm not sure I understand--`SwiftTest` is only ever initialized once below, and no `coverage_mode` is specified. Is the `coverage_mode` parameter ever used?

Comment (harlanhaskins) [inline test/lit.cfg]: It comes from `lit.site.cfg.in` -- CMake fills it in with the appropriate value, and it's used to determine the pre- and post-test behavior

File: test/lit.cfg (hunk 15 above)
Comment (harlanhaskins) [inline test/lit.cfg]: :+1:

File: test/lit.cfg (hunk 16 above)
Comment (modocache) [inline test/lit.cfg]: Yeah, whoops! I deleted my comment too late--I had misread :sweat_smile:

File: utils/profdata_merge_worker.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        self.tmp_dir = tempfile.mkdtemp()
+        self.pid_file_path = os.path.join(self.out_dir, "profdata_merge_worker.pid")
+        self.final_profdata_path = os.path.join(self.out_dir, "swift.profdata")
+        self.remove_files = not no_remove_files
+
+printlock = Lock()
+def printsync(msg):
+    if CONFIG and CONFIG.debug:
+        with printlock:
+            print >>sys.stderr, msg
```
Comment (modocache) [inline utils/profdata_merge_worker.py]: nit-pick: There's been some work done in #90 and #639 to make sure scripts work on both Python 2 and 3. To this end, you should `from __future__ import print_function` at the top of this file, and use the `print()` function here.

Comment (harlanhaskins) [inline utils/profdata_merge_worker.py]: I don't consider that a nitpick at all! :+1:

File: utils/profdata_merge_worker.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+    start.add_argument("--no-remove",
+        action="store_true",
+        help="Don't remove profraw files after merging them.")
+    start.set_defaults(func=start_server)
+
+    stop = subparsers.add_parser("stop")
+    stop.set_defaults(func=stop_server)
+
+    args = parser.parse_args()
+    args.func(args)
```
Comment (modocache) [inline utils/profdata_merge_worker.py]: It would be nice to create a Python module for this, as opposed to a single file. It'd allow you to:

1. Add tests for the module. You could also add the tests to the Swift validation test suite, as in #810.
2. Split the file up. `Config`, `ProfdataMergerProcess`, `ProfdataTCPHandler`, and even the argument parser could each get their own file. Where possible, it would be nice to unit test these classes.

File: utils/profdata_merge_worker.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+            self.report("no files to merge...")
+            return
+        if os.path.exists(self.profdata_path):
+            os.rename(self.profdata_path, self.profdata_tmp_path)
+            self.filename_buffer.append(self.profdata_tmp_path)
+        cleaned_files = ' '.join(pipes.quote(f) for f in self.filename_buffer)
+        # FIXME: This doesn't necessarily always line up with the version
+        #        of clang++ used to build the binaries.
+        llvm_cmd = ("xcrun llvm-profdata merge -o %s %s"
+            % (self.profdata_path, cleaned_files))
```
Comment (modocache) [inline utils/profdata_merge_worker.py]: In addition to not matching the `clang++` used to build the binaries, this command also would not work on Linux. Is this not meant to be cross-platform?

Would it be possible to somehow access `utils/swift_build_support`? `swift_build_support.clang` includes utilities to find the host machine's `clang++`. These might be overridden by specifying `build-script-impl` args, though, so even that's not foolproof...

Comment (harlanhaskins) [inline utils/profdata_merge_worker.py]: We'd love to make it work on Linux, but for right now we're okay with it being OS X-only. This depends on having a suitable version of `compiler-rt` installed alongside clang, which we can guarantee with Xcode.

**Comment by harlanhaskins** [comment]
@modocache I agree that the script would be better off as a module. Let mw know what you think, and thanks so much for the review!

File: utils/profdata_merge/main.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+def printsync(msg, config):
+    if not config.debug:
+        return
+    with printlock:
+        print(msg, file=sys.stderr)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+
+    subparsers = parser.add_subparsers()
```
Comment (modocache) [inline utils/profdata_merge/main.py]: Wow, I had no idea about `add_subparsers()`! Very cool functionality. :+1:

File: utils/profdata_merge/main.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+    start = subparsers.add_parser("start")
+    start.add_argument("-d", "--debug",
+        help="Run in foreground and report status.",
+        action="store_true")
+    start.add_argument("-o", "--output-dir",
+        help="The directory to write the PID file and final profdata file.",
+        default="/tmp")
```
Comment (modocache) [inline utils/profdata_merge/main.py]: Can we use the stdlib's [`tempfile.tempdir`](https://docs.python.org/2/library/tempfile.html#tempfile.tempdir) here instead?

Comment (harlanhaskins) [inline utils/profdata_merge/main.py]: :+1:

File: utils/profdata_merge/process.py [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+from multiprocessing import Process
+from main import printsync
+import pipes
+import os
+import subprocess
+
+class ProfdataMergerProcess(Process):
+    def __init__(self, config, file_queue):
+        Process.__init__(self)
```
Comment (modocache) [inline utils/profdata_merge/process.py]: I think `super(ProfdataMergerProcess, self).__init__()` would be better here. Using `super` mirrors [the sample code in Python's own documentation](https://docs.python.org/2/library/multiprocessing.html#using-a-remote-manager).

Comment (harlanhaskins) [inline utils/profdata_merge/process.py]: Oops, I missed this one. :+1:

File: utils/profdata_merge/main.py (hunk 20 above)
Comment (harlanhaskins) [inline utils/profdata_merge/main.py]: I actually found out about it in `utils/build-script`!

File: utils/profdata_merge_worker.py (hunk 19 above)
Comment (modocache) [inline utils/profdata_merge_worker.py]: Sounds good! Would it be possible to fail gracefully for Linux, then? I imagine users of this script may be confused to suddenly encounter a `subprocess` stack trace where `xcrun` is called.

Comment (harlanhaskins) [inline utils/profdata_merge_worker.py]: It's gonna print an error and exit -- that's now in main.py. :+1:

**Comment by modocache** [comment]
Thank you for the fascinating read, @harlanhaskins. And I've never seen anyone address comments so quickly! :grin: 

My parting thought is that it would be nice if we could verify the behavior of this code without actually building Swift itself--it seems like it would be difficult for someone new to this codebase to verify whether this was working correctly.

**Comment by harlanhaskins** [comment]
@modocache Thanks for the comprehensive review! I really appreciate your input.

The codebase relies really heavily on global out-of-process state, but it's definitely not specific to Swift. This process could be used to merge profile data from instrumented builds of anything, (modulo some hard-coded filenames that explicitly say 'swift'). I could envision a small test runner that compiles a simple C file with profiling and shoots files at the server, then investigates the output, but I don't know if that's on the roadmap for me right away.

Thanks again!

**Comment by harlanhaskins** [comment]
@gribozavr I think this is good to merge. We should probably run the CI once more, though. (@shahmishal, too)

**Comment by shahmishal** [comment]
@swift-ci Please test
